### ISSUE TITLE
cpp:expr role cannot handle initial linebreak

### ISSUE THREAD

**Issue by other-mickk** [sphinx-doc/sphinx#4825]
ping @jakobandersen 

My understanding is that normal rST syntax for interpreted text requires escaping the initial linebreak when there is one:

```rST
.. cpp:var:: template<typename Param> int very_long_template

.. cpp:class:: very_long_arg

Ref: :cpp:any:`\
very_long_template\<
very_long_arg>`.
```

(The second linebreak has no escape which results in extra whitespace once rendered. This can itself be suppressed with its own escape, but I omit it to simplify the demo.) Breaking early like that can be useful when under space constraints, e.g. inside the columns of a table.

The `cpp:expr` role can deal with subsequent linebreaks but not the initial one:

```rst
Ref: :cpp:expr:`\
very_long_template<
very_long_arg>`.
```

Resulting in:

```
path/to/index.rst:11: WARNING: Unparseable C++ expression: '\x00\nvery_long_template<\nvery_long_arg>'
Error when parsing (type) expression.
If expression::
  Error in postfix expression, expected primary expression or type.
  If primary expression:
    Invalid definition: Expected identifier in nested name. [error at 0]
      
    very_long_template<
    very_long_arg>
      ^
  If type:
    Invalid definition: Expected identifier in nested name. [error at 0]
      
    very_long_template<
    very_long_arg>
      ^
If type::
  Invalid definition: Expected identifier in nested name. [error at 0]
    
  very_long_template<
  very_long_arg>
    ^
```

(Incidentally I believe there’s some snafu with unescaping going on in the error report, by which I don't just mean the rawtext but also the caret error location.)

Omitting the initial backslash results in a literal `` :cpp:expr:` `` preceding non-interpreted text. In some cases it's possible to work around the limitation by moving the role to a suffix position.

**Comment by other-mickk** [comment]
> In some cases it's possible to work around the limitation by moving the role to a suffix position.

Although sometimes you run into the same problem, but for a final linebreak:

```rST
This is fine: `very_long_template<
very_long_arg>`:cpp:expr:.

rST syntax error: `foo<bar>
`:cpp:expr:.

Cannot handle escape: `foo<bar>\
`:cpp:expr:.
```

That looks symmetrical to me so I suppose the problem is really about initial or final linebreaks.

**Comment by jakobandersen** [comment]
The first distinction is in directives vs. roles. ``.. cpp:var::`` is a directive, where multiple declarations can be documented by separating them with a line break. Therefore escaping is supported for better formatting, and it is handled independent of domain (https://github.com/sphinx-doc/sphinx/blob/master/sphinx/directives/__init__.py#L75).
For roles I don't know the exact rule for rst-parsing (btw, where is the suffix-position thing documented?). I would guess that if line-breaks are allowed in roles, then no escaping is needed as they only mention one entity anyway.
I believe that almost all roles in domains in Sphinx inherit from the XRefRole so some common text handling is done (https://github.com/sphinx-doc/sphinx/blob/master/sphinx/roles.py#L100). In particular it could be that line-breaks are converted into spaces.
The ``cpp:expr`` role is different, as it is not a single cross-reference, but a rendering of a type/expression. Therefore white-space should be irrelevant for the final rendering.
The error messages with carets indicating the position is a C++ domain-specific thing, which for sure does not handling line-breaks correctly.

I think the ``cpp:expr`` role should simply replace both escaped and unescaped line-breaks with a space before processing (i.e., here: https://github.com/sphinx-doc/sphinx/blob/master/sphinx/domains/cpp.py#L5966).
Similarly I don't think the C++ parser should ever see a line-break from the other roles (though they may be there for the final output).

**Comment by other-mickk** [comment]
> For roles I don't know the exact rule for rst-parsing (btw, where is the suffix-position thing documented?)

I don't know know role syntax either, and whether the escaping is a Docutils or Sphinx thing or not. But in any case, here is the [Docutils interpreted text reference](http://docutils.sourceforge.net/docs/ref/rst/restructuredtext.html#interpreted-text) which documents the suffix variant.

**Comment by jakobandersen** [comment]
This should be fixed in the 1.7 branch now. Escaped line breaks and spaces are stripped, and line breaks are converted to spaces to produce better error messages.
